### DISCUSSION TITLE
JAX.grad

### DISCUSSION THREAD

**Discussion by dvsrikar96** [jax-ml/jax#40544] [Q&A]
I have a loss function that I am using for Bayesian sampling, but the gradients computed using JAX are significantly different from those obtained using a finite-difference approximation in NumPy. Since I have independently verified the finite-difference gradients, I suspect there may be an issue with the JAX implementation or automatic differentiation path. Is there a way to diagnose and rectify this discrepancy while retaining JAX, since the sampling algorithm relies heavily on JAX-based JIT compilation and parallelization?

**Comment by jakevdp** [comment]
Is it possible to provide code that reproduces the issue?

**Comment by Blizzeq** [comment]
Without the loss function it's hard to say which one it is, but these are the usual causes, checked on jax 0.11.2:

1. **float32 on the JAX side.** JAX computes in float32 by default, while the NumPy finite differences are float64. With a step around 1e-6, finite differences taken on a float32 function are mostly rounding noise. On a small test loss I got `[0, 0, 17.17]` from finite differences against a correct `[0.564, -0.299, 18.67]`. To compare like with like, enable x64 before creating any arrays:
   ```python
   jax.config.update("jax_enable_x64", True)
   ```
   If the gradients agree in float64, your loss is probably fine and the sampler is just sensitive to float32 precision.

2. **`jnp.where` around a function that is undefined in the untaken branch.** `jnp.where(x > 0, jnp.sqrt(x), 0.0)` gives a `nan` gradient at `x = 0`, even though the forward value is fine. The fix is to make the untaken branch safe as well:
   ```python
   safe = jnp.where(x > 0, x, 1.0)
   jnp.where(x > 0, jnp.sqrt(safe), 0.0)
   ```
   The same applies to `log`, division and `arccos` inside `where`.

3. **Piecewise operations.** `round`, `floor`, `sign`, `argmax`, comparisons and anything under `lax.stop_gradient` have a zero gradient in JAX. Finite differences can still see a slope there if the step crosses a jump.

To locate the problem, `jax.test_util.check_grads(loss, (x,), order=1, modes=["rev"])` compares autodiff against numerical derivatives and fails with the mismatch. Running it with x64 enabled on pieces of the loss usually points to the term responsible. `jax.config.update("jax_debug_nans", True)` also helps if case 2 is producing NaNs.
